src/fusiontensor/base_interface.jl

```diff
@@ -0,0 +1,165 @@
+  return FusionTensor(
+    adjoint(data_matrix(ft)),
+    dual.(domain_axes(ft)),
+    dual.(codomain_axes(ft)),
+    transpose_mapping(trees_block_mapping(ft)),
+  )
+end
+
+Base.axes(ft::FusionTensor) = (codomain_axes(ft)..., domain_axes(ft)...)
+
+# conj is defined as coefficient wise complex conjugation, without axis dual
+Base.conj(ft::FusionTensor{<:Real}) = ft   # same object for real element type
+function Base.conj(ft::FusionTensor)
+  return FusionTensor(
+    conj(data_matrix(ft)), codomain_axes(ft), domain_axes(ft), trees_block_mapping(ft)
+  )
+end
+
+function Base.copy(ft::FusionTensor)
+  return FusionTensor(
+    copy(data_matrix(ft)),
+    copy.(codomain_axes(ft)),
+    copy.(domain_axes(ft)),
+    copy(trees_block_mapping(ft)),
+  )
+end
+
+function Base.deepcopy(ft::FusionTensor)
+  return FusionTensor(
+    deepcopy(data_matrix(ft)),
+    deepcopy.(codomain_axes(ft)),
+    deepcopy.(domain_axes(ft)),
+    deepcopy(trees_block_mapping(ft)),
+  )
+end
+
+# eachindex is automatically defined for AbstractArray. We do not want it.
+Base.eachindex(::FusionTensor) = error("eachindex not defined for FusionTensor")
+
+function Base.getindex(ft::FusionTensor, f1f2::Tuple{<:SectorFusionTree,<:SectorFusionTree})
+  return ft[f1f2...]
+end
+function Base.getindex(ft::FusionTensor, f1::SectorFusionTree, f2::SectorFusionTree)
+  charge_matrix = data_matrix(ft)[trees_block_mapping(ft)[f1, f2]]
+  return reshape(charge_matrix, charge_block_size(ft, f1, f2))
+end
+
+function Base.setindex!(
+  ft::FusionTensor, a::AbstractArray, f1f2::Tuple{<:SectorFusionTree,<:SectorFusionTree}
+)
+  return setindex!(ft, a, f1f2...)
+end
+function Base.setindex!(
+  ft::FusionTensor, a::AbstractArray, f1::SectorFusionTree, f2::SectorFusionTree
+)
+  return view(ft, f1, f2) .= a
+end
+
+Base.ndims(::FusionTensor{T,N}) where {T,N} = N
+
+Base.permutedims(ft::FusionTensor, args...) = fusiontensor_permutedims(ft, args...)
+
+function Base.similar(ft::FusionTensor)
+  mat = similar(data_matrix(ft))
+  initialize_allowed_sectors!(mat)
+  return FusionTensor(mat, codomain_axes(ft), domain_axes(ft), trees_block_mapping(ft))
+end
+
+function Base.similar(ft::FusionTensor, ::Type{T}) where {T}
+  # fusion trees have Float64 eltype: need compatible type
+  @assert promote_type(T, Float64) === T
+  mat = similar(data_matrix(ft), T)
+  initialize_allowed_sectors!(mat)
+  return FusionTensor(mat, codomain_axes(ft), domain_axes(ft), trees_block_mapping(ft))
+end
+
+function Base.similar(::FusionTensor, ::Type{T}, new_axes::Tuple{<:Tuple,<:Tuple}) where {T}
+  return FusionTensor(T, new_axes[1], new_axes[2])
+end
+
+Base.show(io::IO, ft::FusionTensor) = print(io, "$(ndims(ft))-dim FusionTensor")
+
+function Base.show(io::IO, ::MIME"text/plain", ft::FusionTensor)
+  println(io, "$(ndims(ft))-dim FusionTensor with axes:")
+  for ax in axes(ft)
+    display(ax)
+    println(io)
+  end
+  return nothing
+end
+
+Base.size(ft::FusionTensor) = quantum_dimension.(axes(ft))
```

Comment:
Related to the discussion elsewhere, probably we should redefine `length(::AbstractGradedUnitRange)` to output the number of basis vectors in the vector space as opposed to the number of independent parameters, so then this can be defined as: `Base.size(ft::FusionTensor) = length.(axes(ft))` (though I think we won't need to define this overload since Base defines `size` automatically for `AbstractArray`s as long as you define `axes`).